compiler/AST/iterator.cpp

```diff
@@ -2128,11 +2149,139 @@ static void addLocalsToClassAndRecord(Vec<Symbol*>& locals, FnSymbol* fn,
+  } else {
+    for_formals(formal, fn) {
+      if (formal->hasFlag(FLAG_RETARG)) returnType = formal->type->getValType();
+    }
+  }
+
+  CHPL_ASSERT(returnType);
+  auto recordType = toAggregateType(returnType);
+  CHPL_ASSERT(recordType);
+  CHPL_ASSERT(recordType->aggregateTag == AGGREGATE_RECORD);
+  CHPL_ASSERT(recordType->symbol->hasFlag(FLAG_THUNK_RECORD));
+
+  return recordType;
+}
+
+static void addLocalsToThunkRecord(FnSymbol* fn,
+                                   Vec<Symbol*>& locals,
+                                   SymbolMap& local2field,
+                                   AggregateType* thunkRecord) {
+  int counter = 0;
+  for (auto local : locals) {
+    Symbol* field = createICField(counter, local, NULL, /* isValueField */ false, fn);
+    local2field.put(local, field);
+    thunkRecord->fields.insertAtTail(new DefExpr(field));
+  }
+}
+
+static void replaceLocalsWithThunkFields(FnSymbol* fn,
+                                         Symbol* tr,
+                                         std::vector<BaseAST*>& asts,
+                                         Vec<Symbol*>& locals,
+                                         SymbolMap& local2field) {
+  Vec<SymExpr*> defSet;
+  Vec<SymExpr*> useSet;
+  buildDefUseSets(locals, fn, defSet, useSet);
+
+  for (BaseAST* ast : chpl::expandingIterator(asts)) {
+    auto se = toSymExpr(ast);
+    if (!se || !se->parentSymbol) continue;
+
+    replaceLocalUseOrDefWithFieldRef(se, tr, asts, local2field, defSet, useSet);
+  }
+}
+
+static void populateThunkInvokeFn(FnSymbol* fn, FnSymbol* invokeFn) {
+  auto thunkBody = fn->body->body.last();
+  invokeFn->body->replace(thunkBody->remove());
+
+  // Find the thunk result, transform it into a return at the end of the function.
+  for_alist_backward(expr, invokeFn->body->body) {
+    if (auto call = toCallExpr(expr)) {
+      if (call->isPrimitive(PRIM_THUNK_RESULT)) {
+        auto toReturn = call->get(1);
+        call->remove();
+        insertReturn(invokeFn, toSymExpr(toReturn)->symbol());
+        break;
+      }
+    }
+  }
+}
+
+static void rebuildThunkBuilder(FnSymbol* fn,
+                                Vec<Symbol*>& locals,
+                                SymbolMap& local2field,
+                                AggregateType* thunkRecord) {
+  fn->retSymbol = NULL;
+  for_alist(expr, fn->body->body)
+    expr->remove();
+
+  Symbol* thunk = newTemp("_ir", thunkRecord);
+  fn->insertAtTail(new DefExpr(thunk));
+
+  // Avoids a valgrind warning about uninitialized memory when performing
+  // indirect modification checks on const and default intent arguments
+  if (fWarnUnstable && !fNoConstArgChecks) {
+    fn->insertAtTail(new CallExpr(PRIM_ZERO_VARIABLE, new SymExpr(thunk)));
+  }
+
+  initializeRecordFieldWithArgLocals(fn, thunk, locals, local2field);
+
+  // Return the filled-in iterator record.
+  insertReturn(fn, thunk);
+
+  fn->addFlag(FLAG_INLINE);
+}
+
+void lowerThunk(FnSymbol* fn) {
```

Comment:
I did this. It required making some static functions in `iterator.h` public, since a lot of the logic from `iterator.cpp` is shared. Let me know if you think that's fine.

Reply:
I think that's a step in the right direction certainly. A more ideal solution (that I am not asking for) is to put those functions in a common header (I am not sure what I would name that header myself) to be shared between the iterator and the thunk logic.

No further action needed for my sake, though.